File: common-java/src/main/java/com/common/message/Response.java
package com.common.message;

import java.util.Map;

public class Response {
    private Object data;
    private Status status;

    public Response() {}

    public static Response dataResponse(Object data) {
        Response response = new Response();
        response.setData(data);
        return response;
    }

    public static Response errorResponse(String code, Map<String, Object> messageParams) {
        Response response = new Response();
        Status status = new Status();
        status.setCode(code);
        status.setMessageParams(messageParams);
        response.setStatus(status);
        return response;
    }

    public static Response systemErrorResponse(String code) {
        Response response = new Response();
        Status status = new Status();
        status.setCode(code);
        status.setIsSystemError(true);
        response.setStatus(status);
        return response;
    }

    public boolean hasError() {
        return this.status != null && this.status.getCode() != null && !this.status.getCode().isEmpty();
    }

    public Object getData() { return data; }
    public void setData(Object data) { this.data = data; }

    public Status getStatus() { return status; }
    public void setStatus(Status status) { this.status = status; }
}

File: common-golang/message/response.go
package message

type ParamError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

type Status struct {
	Code          string         `json:"code,omitempty"`
	MessageParams map[string]any `json:"messageParams,omitempty"`
	Source        string         `json:"source,omitempty"`
	Params        []ParamError   `json:"params,omitempty"`
	IsSystemError bool           `json:"isSystemError,omitempty"`
}

// Sanitize strips internal details, safe to return to external clients.
func (s *Status) Sanitize() *Status {
	if s == nil {
		return nil
	}
	safe := &Status{
		Code:          s.Code,
		MessageParams: s.MessageParams,
	}
	if s.IsSystemError {
		safe.Code = "INTERNAL_SERVER_ERROR"
		safe.MessageParams = nil
	}
	return safe
}

type Response struct {
	Data   any     `json:"data,omitempty"`
	Status *Status `json:"status,omitempty"`
}

func DataResponse(data any) *Response {
	return &Response{Data: data}
}

func ErrorResponse(code string, messageParams map[string]any) *Response {
	return &Response{
		Status: &Status{
			Code:          code,
			MessageParams: messageParams,
		},
	}
}

func SystemErrorResponse(code string) *Response {
	return &Response{
		Status: &Status{
			Code:          code,
			IsSystemError: true,
		},
	}
}

func (r *Response) HasError() bool {
	return r != nil && r.Status != nil && r.Status.Code != ""
}
